src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,392 @@
+        )
+
+        now = datetime.now(tzinfo=timezone.utc)
+
+        # create a daily weighted average for the next 24h
+        avg_24h = feature_extractor.avg(
+            now,
+            now + timedelta(hours=24),
+            weights=[0.1, 0.2, 0.3, 0.4]
+        )
+
+        # create a daily average for Thursday March 23 2023
+        th_avg_24h = feature_extractor.avg(datetime(2023,3,23), datetime(2023,3,24))
+        ```
+    """
+
+    def __init__(
+        self,
+        moving_window: MovingWindow,
+        period: timedelta,
+    ) -> None:
+        """
+        Initialize a PeriodicFeatureExtractor object.
+
+        Args:
+            moving_window: The MovingWindow that is used for the average calculation.
+            period: The distance between two succeeding intervals.
+
+        Raises:
+            ValueError: If the MovingWindow size is not a integer multiple of the period.
+        """
+        self._moving_window = moving_window
+        self._buffer = moving_window._buffer
+
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
+        """The sampling_period as integer to use it for indexing of samples."""
+
+        self._period = int(period.total_seconds() / self._sampling_period)
+        """Distance between two succeeding intervals in samples."""
+
+        log.debug("len(moving_window): %i", len(self._moving_window))
+        log.debug("period: %i", self._period)
+
+        if not len(self._moving_window) % self._period == 0:
+            raise ValueError(
+                "The MovingWindow size is not a integer multiple of the period."
+            )
+
+        if self._period - period.total_seconds() / self._sampling_period != 0:
+            log.warning(
+                "The period is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
+
+        log.debug("Initializing PeriodicFeatureExtractor!")
+        log.debug(
+            "Period between two succeeding intervals (in samples): %i",
+            self._period,
+        )
+
+    def _timestamp_to_rel_index(self, timestamp: datetime) -> int:
+        """
+        Get the index of a timestamp relative to the oldest sample in the MovingWindow.
+
+        In other word consider an integer axis where the zero is defined as the
+        oldest element in the MovingWindow. This function returns the index of
+        the given timestamp an this axis.
+
+        Args:
+            timestamp: A timestamp that we want to shift into the window.
+
+        Returns:
+            The index of the timestamp shifted into the MovingWindow.
+        """
+        # align timestamp to the sampling period
+        timestamp = self._buffer.normalize_timestamp(timestamp)
+
+        # distance between the input ts and the ts of oldest know samples (in samples)
+        dist_to_oldest = int(
+            abs((timestamp - self._moving_window.time_bound_oldest).total_seconds())
```

Comment:
Using abs is a bug. Using modulo arithmetic alone is enough and thus it's fine to return a negative value here.